### PR TITLE
add rule for _C_ops.matmul

### PR DESCRIPTION
**Pull request by greenhandF**
add mapping and rule for paddle.C_ops.matmul

### CODE REVIEW

**Review by cangtianhuang — DISMISSED** [review]
LGTM

File: tester/accuracy.py [hunk 1]
```diff
@@ -377,7 +377,7 @@ def compare_paddle_and_torch(paddle_tensor, torch_tensor, idx=0) -> bool:
                     paddle_item is None
                     or (
                         isinstance(paddle_item, paddle.Tensor)
-                        and not (paddle_item._is_initialized() or paddle_item.numel() == 0)
+                        and (not paddle_item._is_initialized() or paddle_item.numel() == 0)
```
Comment (cangtianhuang) [inline tester/accuracy.py]: 这样会影响合法的 0size tensor 正常判断（0size tensor 的 numel 为 0，但是有数据类型与形状），是否能改成：

```suggestion
                        and not paddle_item._is_initialized()
```

File: tester/accuracy.py [hunk 2]
```diff
@@ -511,7 +511,7 @@ def compare_paddle_and_torch(paddle_tensor, torch_tensor, idx=0) -> bool:
                         paddle_item is None
                         or (
                             isinstance(paddle_item, paddle.Tensor)
-                            and not (paddle_item._is_initialized() or paddle_item.numel() == 0)
+                            and (not paddle_item._is_initialized() or paddle_item.numel() == 0)
```
Comment (cangtianhuang) [inline tester/accuracy.py]: 同上

File: tester/paddle_to_torch/rules.py [hunk 3]
```diff
@@ -7368,6 +7448,87 @@ def apply(self, paddle_api: str) -> ConvertResult:
...
+    #   probs_grad = empty({outer}, fp32)   # 始终新分配
+    # 当任一输入 numel==0 时，kernel 不执行，直接返回上述（已分配但未写入的）buffer。
+    # 因此 inplace=True 时 do1/o2_s 保留 o1/do2_s 的原值；非 inplace 时为未初始化值。
+    if o1.numel() == 0 or do2_s.numel() == 0 or probs.numel() == 0:
+        if inplace_flag:
+            do1_e  = o1.clone()
+            o2_s_e = do2_s.clone()
+        else:
+            do1_e  = torch.zeros_like(o1)
+            o2_s_e = torch.zeros_like(do2_s)
```
Comment (cangtianhuang) [inline tester/paddle_to_torch/rules.py:7484]: 可能 empty_like 更贴切～但是都行

File: tester/paddle_to_torch/rules.py [hunk 4]
```diff
@@ -7368,6 +7448,87 @@ def apply(self, paddle_api: str) -> ConvertResult:
...
+            o2_val   = silu_lhs * rhs_c
+            do2_val  = do2_c * prob_c
+            x0g = do2_val * rhs_c * sig * (1.0 + lhs_c - silu_lhs)
+            x1g = do2_val * silu_lhs
+            pg_out[s:e] = (do2_c * o2_val).sum(dim=-1)
+            # o2_s 写入需在 do1 之前完成时使用 do2_2d 切片；这里 o2_s 与 do1
+            # 来自不同 buffer（即使 inplace 也是 do2_s vs o1），互不影响。
+            o2_s_out[s:e] = (o2_val * prob_c).to(do2_dtype)
+            do1_out[s:e, :H] = x0g.to(o1_dtype)
+            do1_out[s:e, H:] = x1g.to(o1_dtype)
```
Comment (cangtianhuang) [inline tester/paddle_to_torch/rules.py]: 这处切片赋值会对 fp32 的叶子节点进行原地写入，torch 会报错，考虑加 `torch.no_grad()` 包裹

可以复现一下：
`paddle._C_ops._run_custom_op("fused_swiglu_probs_bwd", Tensor([2, 4],"float32"), Tensor([2, 2],"float32"), Tensor([2, 1],"float32"), True, )`

**Review by cangtianhuang — CHANGES_REQUESTED** [review]
修改后要统一复测一下原有配置，在pr描述里贴上测试结果，要看精度没有大问题～

**Comment by greenhandF** [comment]
apitest
所有paddle_only能够通过的test精度比较都没有问题，共3个算子共23个case

**Review by cangtianhuang — DISMISSED** [review]
LGTM

**Review by wanghuancoder — DISMISSED** [review]
LGTM
